src/net_processing.cpp

```diff
@@ -2444,36 +2443,36 @@ bool ProcessMessages(CNode* pfrom, CConnman& connman)
         ProcessGetData(pfrom, chainparams.GetConsensus(), connman);
 
     // this maintains the order of responses
-    if (!pfrom->vRecvGetData.empty()) return fOk;
-
-    std::deque<CNetMessage>::iterator it = pfrom->vRecvMsg.begin();
-    while (!pfrom->fDisconnect && it != pfrom->vRecvMsg.end()) {
-        // Don't bother if send buffer is too full to respond anyway
-        if (pfrom->nSendSize >= nMaxSendBufferSize)
-            break;
-
-        // get next message
-        CNetMessage& msg = *it;
-
-        //if (fDebug)
-        //    LogPrintf("%s(message %u msgsz, %u bytes, complete:%s)\n", __func__,
-        //            msg.hdr.nMessageSize, msg.vRecv.size(),
-        //            msg.complete() ? "Y" : "N");
-
-        // end, if an incomplete message is found
-        if (!msg.complete())
-            break;
-
-        // at this point, any failure means we can delete the current message
-        it++;
+    if (!pfrom->vRecvGetData.empty()) {
+        connman.WakeMessageHandler();
+        return fOk;
+    }
 
-        // Scan for message start
-        if (memcmp(msg.hdr.pchMessageStart, chainparams.MessageStart(), CMessageHeader::MESSAGE_START_SIZE) != 0) {
-            LogPrintf("PROCESSMESSAGE: INVALID MESSAGESTART %s peer=%d\n", SanitizeString(msg.hdr.GetCommand()), pfrom->id);
-            fOk = false;
+    while (!pfrom->fDisconnect) {
+        std::vector<CNetMessage> vMsg;
+        if (pfrom->nSendSize >= nMaxSendBufferSize) {
+            connman.WakeMessageHandler();
             break;
         }
+        {
+            TRY_LOCK(pfrom->cs_vRecvMsg, lockRecv);
+            if (lockRecv)
+            {
+                if (!pfrom->vRecvMsg.empty()) {
+                    if (!pfrom->vRecvMsg.front().complete()) break;
+                    // Move the message.
+                    vMsg.push_back(std::move(pfrom->vRecvMsg.front()));
+                    if (pfrom->fDisconnect) break;
+                    pfrom->vRecvMsg.pop_front();
+                }
+            } else {
+                connman.WakeMessageHandler();
+            }
+        }
```

Comment:
Looping here to process all outstanding messages might not be desirable because it could let a single peer monopolize processing. Any thoughts on that?